llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -12070,6 +12070,17 @@ SDValue DAGCombiner::visitVSELECT(SDNode *N) {
   if (SDValue F = extractBooleanFlip(N0, DAG, TLI, false))
     return DAG.getSelect(DL, VT, F, N2, N1);
 
+  // select (sext m), (add X, C), X --> (add X, (and C, (sext m))))
+  if (N1.getOpcode() == ISD::ADD && N1.getOperand(0) == N2 && N1->hasOneUse() &&
+      DAG.isConstantIntBuildVectorOrConstantInt(N1.getOperand(1)) &&
+      N0.getScalarValueSizeInBits() == N1.getScalarValueSizeInBits() &&
+      TLI.getBooleanContents(N0.getValueType()) ==
+          TargetLowering::ZeroOrNegativeOneBooleanContent) {
```

Comment:
I wonder if this can prevent for AVX512. I think we don't want this on AVX512 because we have masked instruction.

Reply:
The `N0.getScalarValueSizeInBits() == N1.getScalarValueSizeInBits()` check shouldn't be true for AVX512 since the mask won't be promoted.